addons/binding/org.openhab.binding.denonmarantz/src/main/java/org/openhab/binding/denonmarantz/handler/DenonMarantzHandler.java

```diff
@@ -0,0 +1,208 @@
+    }
+
+    private void createConnection() {
+        configuration = getConfigAs(DenonMarantzConfiguration.class);
+        if (connector != null) {
+            connector.dispose();
+        }
+        connector = new DenonMarantzConnector(configuration, this, scheduler);
+        connector.connect();
+    }
+
+    private void addZoneChannels() {
+        // only add when this is a new Thing instance (and doesn't have the 'zones' property set yet).
+
+        if (this.getThing().getProperties().containsKey("zones")) {
+            return;
+        }
+        if (configuration.getZoneCount() > 1) {
+            logger.debug("Adding zone channels");
+            // add channels for the secondary zones
+            ThingBuilder thingBuilder = editThing();
+            CopyOnWriteArrayList<Channel> channels = new CopyOnWriteArrayList<Channel>(this.getThing().getChannels());
+
+            for (Entry<ChannelTypeUID, String> entry : DenonMarantzBindingConstants.ZONE2_CHANNEL_TYPES.entrySet()) {
+                ChannelType type = TypeResolver.resolve(entry.getKey());
+                Channel channel = ChannelBuilder
+                        .create(new ChannelUID(this.getThing().getUID(), entry.getValue()), type.getItemType())
+                        .withType(entry.getKey()).build();
+                channels.add(channel);
+            }
+
+            if (configuration.getZoneCount() > 2) {
+                // add channels for zone 3 (more zones currently not supported)
+                for (Entry<ChannelTypeUID, String> entry : DenonMarantzBindingConstants.ZONE3_CHANNEL_TYPES
+                        .entrySet()) {
+                    ChannelType type = TypeResolver.resolve(entry.getKey());
+                    Channel channel = ChannelBuilder
+                            .create(new ChannelUID(this.getThing().getUID(), entry.getValue()), type.getItemType())
+                            .withType(entry.getKey()).build();
+                    channels.add(channel);
+                }
+            }
+
+            thingBuilder.withChannels(channels);
+            Map<String, String> properties = editProperties();
+            properties.put("zones", configuration.getZoneCount().toString());
+            thingBuilder.withProperties(properties);
+            updateThing(thingBuilder.build());
+        }
+    }
+
+    @Override
+    public void dispose() {
+        connector.dispose();
+        super.dispose();
+    }
+
+    @Override
+    public void channelLinked(ChannelUID channelUID) {
+        super.channelLinked(channelUID);
+        String channelID = channelUID.getId();
+        if (isLinked(channelID)) {
+            updateState(channelID, connector.getState().getStateForChannelID(channelID));
+        }
+    }
+
+    @Override
+    public void stateChanged(String channelID, State state) {
+        logger.debug("Recieved state {} for channelID {}", state, channelID);
+        if (this.getThing().getStatus() != ThingStatus.ONLINE) {
```

Comment:
Removed this.
Did apply similar check for setting the Thing to OFFLINE only once though. As it can be set offline with different descriptions it was causing a flood in the event log in some circumstances (see https://community.openhab.org/t/denon-binding-warn-log-messages/33024/12)

Reply:
Re-added it after I saw the events.log getting filled with `'denonmarantz:avr:15f35735' updated: ONLINE`. If any Channel state changes, it goes through this function.
We only want to notify the system about the Thing's online status when it actually changed to ONLINE.